### PR TITLE
Feature/49 Init command after run command has started

### PR DESCRIPTION
**Pull request by puengel**
This PR adds `init` and `initOnce`  to the `run:` directive. Bot scripts are executed by sh mvdan.cc/sh
Nix support is missing. Could be added in another PR. Tracked here #97 

### Todo:
- [x] add init field
- [x] run init commands
- [x] make init comands cancellable
- [x] added tests
- [x] option to run init only once 
- [ ] ~test on medium size project~

### CODE REVIEW

File: bobrun/wrap.go [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+			s.Split(bufio.ScanLines)
+
+			go func() {
+				for s.Scan() {
+					err := s.Err()
+					if err != nil {
+						return
+					}
+
+					// FIXME: why does printing to Commands stdout not work?
```
Comment (Equanox) [inline bobrun/wrap.go:205]: @rdnt i have several FIXME's related to piping output to the status tab.. Does it make sense to look over them in this PR, or shall we create a entire Issue to look into the topic (i know you already did a lot to make it work).

Comment (Equanox) [inline bobrun/wrap.go:205]: @rdnt i have several FIXME's related to piping output to the status tab.. Does it make sense to look over them in this PR, or shall we create a entire Issue to look into the topic (i know you already did a lot to make it work).

File: bobrun/map.go [hunk 2]
```diff
@@ -26,3 +29,22 @@ func (rm RunMap) String() string {
...
 	return description.String()
 }
+
+// Sanitize run map and write filtered & sanitized
+// propertys from dirty members to plain (e.g. dirtyInit -> init)
+func (rm RunMap) Sanitize() (err error) {
+	defer errz.Recover(&err)
+
+	for key, task := range rm {
+
```
Comment (zuzuleinen) [inline bobrun/map.go:39]: Here I would get rid of empty temp variables sanitizedInits, sanitizedInitOnce:

```go
	for key, task := range rm {
		task.init = multilinecmd.Split(task.InitDirty)
		task.initOnce = multilinecmd.Split(task.InitOnceDirty)

		rm[key] = task
	}
```

File: bobrun/run.go [hunk 3]
```diff
@@ -20,6 +21,15 @@ type Run struct {
 	// DependsOn run or build tasks
 	DependsOn []string
 
+	InitDirty string `yaml:"init"`
```
Comment (zuzuleinen) [inline bobrun/run.go:24]: Maybe add a comment for InitDirty similar to InitOnceDirty

File: bobrun/run.go [hunk 4]
```diff
@@ -49,24 +59,33 @@ func New() *Run {
...
 	}
 	return r
 }
 
-// Run creates run cmds and return a channel to ctl it.
-// To shutdown a Run() use a cancable context.
-func (r *Run) Run(ctx context.Context) (rc ctl.Command, _ error) {
-	// fmt.Printf("Creating control for run task [%s]\n", r.name)
+// Command creates a run cmd and returns a Command interface to control it.
+// To shutdown a Command() use a cancable context.
```
Comment (zuzuleinen) [inline bobrun/run.go:71]: typo in cancable

File: bobrun/wrap.go [hunk 5]
```diff
@@ -0,0 +1,268 @@
...
+	"github.com/benchkram/errz"
+	"github.com/logrusorgru/aurora"
+	"mvdan.cc/sh/expand"
+	"mvdan.cc/sh/interp"
+	"mvdan.cc/sh/syntax"
+)
+
+// Wrapper wraps a run-task to provide init functionality executed after
+// the task started.
+type Wrapper struct {
```
Comment (zuzuleinen) [inline bobrun/wrap.go:24]: Maybe Wrapper is too general for name? I would change it to `InitWrapper` or `WithInit`.

**Review by zuzuleinen — APPROVED** [review]
approved with minor comments regarding code style & typos

File: bob/bobfile/bobfile.go [hunk 6]
```diff
@@ -256,7 +256,13 @@ func BobfileRead(dir string) (_ *Bobfile, err error) {
 	err = b.Validate()
 	errz.Fatal(err)
 
-	return b, b.BTasks.Sanitize()
+	err = b.BTasks.Sanitize()
+	errz.Fatal(err)
+
+	err = b.RTasks.Sanitize()
```
[the remark this answers is not in the document]
Comment (Equanox) [reply bob/bobfile/bobfile.go:262]: Once for **B**uildTasks and once for **R**unTasks

**Pull request merged**